Tutorial: Get started with k6 (#1192)
grafana/k6-docs: src/data/markdown/docs/05 Examples/02 Tutorials/01 Get started with k6/400 Reuse and re-run tests.md

---
title: Reuse and re-run tests 
excerpt: Modularize your k6 test logic and workload configuration. 
---

> In the previous tutorials, you designed k6 scripts to assert for performance and to make it easy to compare results.

Reviewer: ```suggestion
In the previous tutorials, you designed k6 scripts to assert performance and make comparing results easy.
```

In this tutorial, learn how to:
- Modularize test scripts into reusable components
- Dynamically configure scripts with environment variables

## Example script

For fun, let's combine the scripts from the previous tutorials.

> Use logic of the `multiple-flows.js` test with the thresholds and scenario of the `api-test.js` (feel free to add more checks, requests, groups, and so on).

Reviewer: ```suggestion
Use the logic of the `multiple-flows.js` test with the thresholds and scenario of the `api-test.js` (feel free to add more checks, requests, groups, and so on).
```

Take note of the features of this script:

> - The `default` function has two groups, `Contacts flow` and `Coinflip game`

Reviewer: ```suggestion
- The `default` function has two groups, `Contacts flow`, and `Coinflip game`
```

  coinflip(baseUrl);
}
```

```javascript
export const thresholdsSettings = { 
  http_req_failed: [{ threshold: "rate<0.01", abortOnFail: true }],
  http_req_duration: ["p(99)<1000"],
}

export const breakingWorkload = {
  executor: 'ramping-vus',
  stages: [
    { duration: '10s', target: 20 },
    { duration: '50s', target: 20 },
    { duration: '50s', target: 40 },
    { duration: '50s', target: 60 },
    { duration: '50s', target: 80 },
    { duration: '50s', target: 100 },
    { duration: '50s', target: 120 },
    { duration: '50s', target: 140 },
    //....
  ],
} 
```

</CodeGroup>

Notice the length of this final script and compare it with the script at the beginning of this page.
Though the final execution is the same, it's half the size and more readable.

Besides shortness, this modularity lets you compose scripts from many parts, or dynamically configure scripts at run time.

## Mix and match logic

With modularized configuration and logic, you can mix and match logic.
An easy way to configure this is through [environment variables](/using-k6/environment-variables).

Change `main.js` and `config.js` so that it:
- _By default_ runs a smoke test with 5 iterations

> - With the right enviroment variable value, runs a breaking test

Reviewer: ```suggestion
- With the right environment variable value, runs a breaking test
```

   
  <CodeGroup labels={["main.js"]} lineNumbers={[false]} showCopyButton={[true]} heightTogglers={[true]}>

  ```javascript

  import { coinflip } from "./coinflip.js";
  import { contacts } from "./contacts.js";
  import { thresholdsSettings, breakingWorkload, smokeWorkload } from "./config.js";

  export const options = {
    scenarios: {
      my_scenario: __ENV.WORKLOAD === 'breaking' ? breakingWorkload : smokeWorkload
    },
    thresholds: thresholdsSettings
  };

  const baseUrl = "https://test.k6.io";

  export default function () {
    contacts(baseUrl);
    coinflip(baseUrl);
  }

  ```
  </CodeGroup>
  
1. Run the script with and without the `-e` flag.

   - What happens when you run `k6 run main.js`?
   - What about `k6 run main.js -e WORKLOAD=breaking`?

This was a simple example to showcase how you can modularize a test. 
As your test suite grows and more people are involved in performance testing, your modularization strategy becomes essential to building and maintaining an efficient testing suite.
  
## Next steps

Now you've seen examples to write tests, assert for performance, filter results, and modularize scripts.

> Notice how the tests progress in complexity: from single endpoints to holistic tests, from small loads to large loads, and from single-tests to reusable modules. These progressions are typical in testing, with the next step being to automate. It might be impractical to automate a tutorial, but if you are interested,

Reviewer: ```suggestion
Notice how the tests progress in complexity: from single endpoints to holistic tests, from small to large loads, and from single tests to reusable modules. These progressions are typical in testing, with the next step being to automate. It might be impractical to automate a tutorial, but if you are interested,
```
